### PR TITLE
feat(controller): emit gateway Envoy telemetry from operator-relayed OTel env

### PR DESCRIPTION
**Pull request by pilartomas**
## What

Makes the **agent gateway (Envoy proxy)** emit its own OpenTelemetry. The gateway is the one component that can neither host an in-process SDK (#2317/#2323 style) nor be reached by zero-code auto-instrumentation — it is a C++ data plane — so its telemetry is configured **natively in the controller-rendered bootstrap**. Previously the bootstrap had no tracing, access-log, or stats config at all.

## How it's wired (same env as every platform service)

The exporter target has a dedicated knob decoupled from the controller SDK's env: under `clickstack.enabled` the chart sets `PLATFORM_GATEWAY_OTLP_ENDPOINT`/`_PROTOCOL` pointing gateways at the collector's **gRPC** port (Envoy's stats sink speaks nothing else; the controller's own SDK is OTLP/HTTP-only, so one shared endpoint could never serve both). Without the override the controller falls back to its inherited `OTEL_*` (the BYO case). It relays the effective `OTEL_*` env onto each gateway pod (dropping identity vars and the `*_HEADERS` family) and resolves endpoint/protocol + `OTEL_TRACES_SAMPLER[_ARG]` at render time — Envoy reads none of these natively. With no endpoint present, gateways emit nothing and behave exactly as before.

Three signals when active:
- **Tracing** — OTel tracer on the outer egress listener only. Collector-bound CONNECTs (the agents' own telemetry pushes) ride a dedicated zero-sampled route so the pipeline never traces itself.
- **Access logs** — structured JSON to stdout **and** exported over OTLP to the collector (grpc or http), so gateway logs land in the backend beside every other platform service's logs; the agent-telemetry transit chain gets an **error-only** log so delivery failures are visible.
- **Metrics** — Envoy stats over an OTLP/gRPC sink (admin interface stays disabled). Active under bundled defaults via the gRPC exporter env; an OTLP/HTTP-only BYO collector gets traces + logs but no stats (the sink has no HTTP transport).

Distinct from the #2030 transit chain: that chain forwards the **agent's** telemetry with trusted attribution (`platform.agent.id`); this PR is the gateway's **own** telemetry, exported directly (cluster `otel_export` vs the transit `otel_collector`). Gateway identity rides as `platform.gateway.id` — platform-namespaced because agents can forge `agent.*` resource attributes, and the collector sanitizes only `platform.agent.id`.

## Security invariants (credential-injection MITM proxy)

- Access logs never name the `Authorization` header and strip the query string from logged paths (`REQ_WITHOUT_QUERY`), so query-param-injected credentials can't leak.
- **No trace spans on the credential-injection chains** — post-interception the `:path` can carry a query-param credential and Envoy has no span-tag query stripper. Those egress decisions are already on the api-server's ext_authz audit trail.
- Shared `service.name` across gateways with a bounded `platform.gateway.id` resource attribute (cardinality doesn't scale with agent count).
- Internal trace context (`traceparent`/`tracestate`) is stripped before reaching external upstreams; kept on the internal harness route so spans link to the api-server.
- Collector auth is **transport-level** (ambient ztunnel mTLS + the collector AuthorizationPolicy, which admits the agent namespace); `OTEL_EXPORTER_OTLP_HEADERS` is deliberately not relayed.

## Known limits / follow-up decisions

- Only the base `OTEL_EXPORTER_OTLP_ENDPOINT` (or the gateway override) drives gateway telemetry; per-signal-only configurations instrument the controller but leave gateways dark (documented in logging.md).
- The chart sets no `OTEL_TRACES_SAMPLER`, so gateway egress traces sample at 100% by default. Fine at current scale; a chart-side sampler knob is a possible follow-up.
- Worth one `--mode validate` against the in-cluster Envoy binary before relying on the OTLP/HTTP and sampling paths, and an empirical check that `request_headers_to_remove: [traceparent]` on the plain-HTTP fallthrough isn't undone by router trace-context re-injection.

## Testing / validation

- Controller unit tests cover endpoint+protocol parsing, sampling translation, off-by-default, gRPC vs OTLP/HTTP rendering, credential-redaction invariants, coexistence with the transit chain (unique cluster names — a duplicate is a fatal Envoy config error), leaf-cert gating for the OTel-only combination, the zero-sampled collector route, and the transit error-log.
- `go build`/`vet`/`gofmt`, `helm lint` pass.
- The gRPC render was previously validated against the real `envoyproxy/envoy:distroless-v1.37.2` binary (`--mode validate` → `configuration OK`); the rebased combined render is pinned by a YAML-parse + uniqueness test.

### CODE REVIEW

File: packages/controller/pkg/reconciler/envoy.go [hunk 1]
```diff
@@ -1242,4 +1488,39 @@ func envoyContainer(cfg *config.Config, secrets []corev1.Secret) corev1.Containe
...
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	env := make([]corev1.EnvVar, 0, len(keys)+1)
+	for _, k := range keys {
+		env = append(env, corev1.EnvVar{Name: k, Value: cfg.OTelEnv[k]})
+	}
+	env = append(env, corev1.EnvVar{
+		Name:  "OTEL_RESOURCE_ATTRIBUTES",
+		Value: fmt.Sprintf("agent.id=%s,k8s.namespace.name=%s", instanceName, cfg.Namespace),
```
Comment (dam-code-guardian) [inline packages/controller/pkg/reconciler/envoy.go]: 🟢 **Suggestion:** This sets `OTEL_RESOURCE_ATTRIBUTES` fresh, fully replacing the controller's relayed value. The `agent.id`/namespace override is the right call, but it also drops any platform-wide attributes the Operator may have parked in the inherited `OTEL_RESOURCE_ATTRIBUTES` (e.g. `k8s.cluster.name`, deployment environment). Consider merging the inherited attributes with the gateway-specific ones rather than overwriting, so cluster/deployment context still rides on gateway telemetry. Low impact and non-blocking.

File: packages/controller/pkg/config/config.go [hunk 2]
```diff
@@ -75,6 +87,138 @@ type Config struct {
...
+	return ok
+}
+
+// OTelExporter parses the operator-injected OTEL_EXPORTER_OTLP_ENDPOINT and
+// OTEL_EXPORTER_OTLP_PROTOCOL into the exporter the gateway Envoy bootstrap
+// targets. ok is false when no endpoint was injected. Per the OTLP spec the
+// endpoint is a URL; a bare host[:port] is tolerated. Port defaults to the OTLP
+// convention for the transport (4317 gRPC, 4318 HTTP) when the URL omits it.
+func (c *Config) OTelExporter() (OTLPExporter, bool) {
+	raw := strings.TrimSpace(c.OTelEnv["OTEL_EXPORTER_OTLP_ENDPOINT"])
```
Comment (dam-code-guardian) [inline packages/controller/pkg/config/config.go]: 🟢 **Suggestion:** The parsed exporter discards any path component of `OTEL_EXPORTER_OTLP_ENDPOINT` (only host/port/scheme are kept). For OTLP/HTTP the per-signal URI is later rebuilt as `scheme://host:port/v1/traces`, so a path-prefixed endpoint like `https://collector/otlp` would lose the `/otlp` base the OTLP spec says should prefix the signal path. Unlikely given the in-cluster operator topology (bare `host:4317`/`:4318`), but worth a guard or a doc note if path-prefixed collector endpoints could ever be injected.

**Review by dam-code-guardian — APPROVED** [review]
🛡️ **DAM** — ✅ Code Review @ `0ca8973`

## PR #1810: feat(controller): emit gateway Envoy telemetry from operator-relayed OTel env
**Author:** pilartomas | **Branch:** feat/gateway-otel-ready → main | **Changes:** +765 −15 (7 files)

### Summary
Makes the agent gateway (Envoy) OpenTelemetry-ready by configuring tracing, access logs, and stats natively in the controller-rendered bootstrap, sourced from the `OTEL_*` environment the OpenTelemetry Operator injects (off by default when no endpoint is injected). The controller parses the OTLP endpoint/protocol/sampler env into a template-facing view and relays the env onto each gateway pod. Credential-safety is the headline invariant: no spans on the credential-injection chains, `REQ_WITHOUT_QUERY` on logged paths, no `Authorization` operator, and internal trace context stripped before external upstreams.

### Findings
- 🟢 **Suggestion:** `OTEL_RESOURCE_ATTRIBUTES` is set fresh on the gateway, fully replacing the controller's relayed value (`packages/controller/pkg/reconciler/envoy.go:1523`). The `agent.id`/namespace override is correct, but this also drops any platform-wide attributes the Operator may have parked in `OTEL_RESOURCE_ATTRIBUTES` (e.g. `k8s.cluster.name`, deployment environment). Consider merging the inherited attributes with the gateway-specific ones rather than overwriting, so cluster/deployment context still rides on gateway telemetry. Low impact — most operator setups carry such context in `OTEL_SERVICE_NAME` / separate vars, and `OTEL_SERVICE_NAME` is intentionally not relayed.
- 🟢 **Suggestion:** `OTelExporter()` discards any path component of `OTEL_EXPORTER_OTLP_ENDPOINT` (`packages/controller/pkg/config/config.go:135`). For OTLP/HTTP the per-signal URI is rebuilt as `scheme://host:port/v1/traces`, so a path-prefixed endpoint like `https://collector/otlp` would lose the `/otlp` base that the OTLP spec says should prefix the signal path. Unlikely in the in-cluster operator topology (bare `host:4317`/`:4318`), but worth a guard or a doc note if path-prefixed collector endpoints are ever in scope.
- ✅ **Looks good:** Default-off gating is clean — the zero-value `envoyTelemetryView` renders no telemetry blocks and `gatewayOTelEnv` returns nil, so an uninstrumented platform is byte-for-byte unchanged (pinned by `TestRenderEnvoyBootstrap_TelemetryOffWithoutEndpoint` / `TestEnvoyContainer_NoOTelEnvWhenDisabled`).
- ✅ **Looks good:** Credential-redaction invariants are each backed by a focused render test — tracing provider appears exactly once (outer listener only), `REQ_WITHOUT_QUERY` is used and `%REQ(:PATH)%`/`req(authorization)` are asserted absent, and `traceparent`/`tracestate` are stripped on external egress.
- ✅ **Looks good:** Protocol/transport handling is coherent — gRPC vs HTTP selects http2 vs HTTP/1.1 cluster shape, metrics are gated on `exp.GRPC` (Envoy's stats sink is gRPC-only), and https → upstream TLS with SNI; all covered by tests. Env keys are sorted for a stable pod spec across reconciles.

### Documentation Check (doc-drift)
**Verdict: aligned**

✅ No documentation drift detected. The gateway behavior change is reflected in `docs/architecture/logging.md` via a new "Gateway telemetry" section (relay model, three signals, and the four credential-redaction invariants); its `Last verified:` header is bumped to the PR date (2026-06-19 → 2026-06-24); and the added prose stays within the documentation guidelines — no ADR references and no volatile package/path leakage (only OTel/W3C/Envoy protocol-contract identifiers).

### Verdict
APPROVE — Well-structured, default-off, and thoroughly tested telemetry wiring with carefully-preserved credential-safety invariants; the two suggestions are non-blocking edge cases.

---
_Review by [DAM](https://github.com/dam-agents/dam) · automated code guardian_

**Review by dam-code-guardian — APPROVED** [review]
🛡️ **DAM** — ✅ Code Review @ `f48dd55`

## PR #1810: feat(controller): emit gateway Envoy telemetry from operator-relayed OTel env
**Author:** pilartomas | **Branch:** feat/gateway-otel-ready → main | **Changes:** +775 −17 (7 files)

### Summary
Makes the agent gateway (Envoy) OpenTelemetry-ready by configuring tracing, access logs, and stats natively in the controller-rendered bootstrap, sourced from the `OTEL_*` environment the OpenTelemetry Operator injects (off by default when no endpoint is injected), with credential-safety as the headline invariant. The new commit since the last review tightens that invariant: the OTLP `*_HEADERS` env family is no longer relayed onto gateway pods — collector auth is transport-level (mesh mTLS), so relaying header-based auth would be inert for Envoy and would needlessly spread a collector credential.

### Changes since last review
Previous HEAD: `0ca8973` (2026-06-24T09:58:38Z) — verdict APPROVE

- 🆕 **New:** OTLP `*_HEADERS` env vars are now dropped from the gateway relay (`packages/controller/pkg/reconciler/envoy.go:1514`) — security-positive, and backed by `TestEnvoyContainer_RelaysOTelEnvWithGatewayIdentity` asserting `OTEL_EXPORTER_OTLP_HEADERS` (carrying a bearer token) does not ride onto the gateway. Doc updated in `docs/architecture/logging.md` to describe transport-level collector auth.
- 🔁 **Still present:** `OTEL_RESOURCE_ATTRIBUTES` is set fresh on the gateway, fully replacing the controller's relayed value (`packages/controller/pkg/reconciler/envoy.go`) — dropping any platform-wide attributes the Operator parked there. Non-blocking; carried over from the previous review.
- 🔁 **Still present:** `OTelExporter()` discards any path component of `OTEL_EXPORTER_OTLP_ENDPOINT` (`packages/controller/pkg/config/config.go:135`), so a path-prefixed OTLP/HTTP endpoint would lose its base path. Non-blocking; carried over.

### Findings
- 🆕 **Looks good (new):** Dropping the `*_HEADERS` family via `strings.HasSuffix(k, "HEADERS")` (`packages/controller/pkg/reconciler/envoy.go:1514`) covers the generic and per-signal header vars (`OTEL_EXPORTER_OTLP_HEADERS`, `..._TRACES_HEADERS`, etc.) in one guard; the suffix match is correct since every OTel `*HEADERS` var is collector-auth metadata Envoy can't consume from env. Tested.
- 🟡 **Warning (carryover, summary-only):** `OTEL_RESOURCE_ATTRIBUTES` override drops inherited platform-wide attributes — consider merging rather than overwriting. (`packages/controller/pkg/reconciler/envoy.go`)
- 🟢 **Suggestion (carryover, summary-only):** Path component of `OTEL_EXPORTER_OTLP_ENDPOINT` is discarded; add a guard or doc note if path-prefixed collector endpoints are ever in scope. (`packages/controller/pkg/config/config.go:135`)

### Documentation Check (doc-drift)
**Verdict: aligned**

✅ No documentation drift detected. The new commit's behavior change (transport-level collector auth, `*_HEADERS` not relayed) is reflected in `docs/architecture/logging.md`'s "Gateway telemetry" section; `Last verified:` is at the PR date (2026-06-24); no ADR references and no volatile package/path leakage (only OTel/Envoy protocol-contract identifiers).

### Verdict
APPROVE — The follow-up commit is a clean, well-tested security tightening of an already-solid telemetry wiring; the two carried-over suggestions remain non-blocking edge cases.

---
_Review by [DAM](https://github.com/dam-agents/dam) · automated code guardian_

**Comment by willallstet** [comment]
📝 Docs reviewed for this change — no documentation update required (internal observability wiring (gateway Envoy telemetry from relayed OTEL_* env); no user-facing knobs or behavior.)

**Review by dam-code-guardian — APPROVED** [review]
🛡️ **DAM** — ✅ Code Review @ `1080095`

## PR #1810: feat(controller): emit gateway Envoy telemetry from operator-relayed OTel env
**Author:** pilartomas | **Branch:** feat/gateway-otel-ready → main | **Changes:** +775 −15 (7 files)

### Summary
Makes the agent gateway (Envoy) OpenTelemetry-ready: the controller relays its own inherited `OTEL_*` env onto each gateway pod and renders native Envoy tracing / access-log / metrics config in the bootstrap (the chart owns no collector config — that stays the Operator's concern). Off by default; credential bytes are kept out of spans and logs by design. This re-review covers one new commit since `f48dd55`. _(@willallstet's "docs reviewed — no update required" comment noted; the PR does update `logging.md` regardless. No findings suppressed; no PR-local overrides on file.)_

### Changes since last review
Previous HEAD: `f48dd55` (2026-06-24T12:57:01Z) — verdict APPROVE

- 🆕 **New — ✅:** `feat(controller): keep gateway collector auth transport-level, drop header relay` (`envoy.go:1516`) — `gatewayOTelEnv` now skips any inherited env key ending in `HEADERS` (`strings.HasSuffix(k, "HEADERS")`), so the `OTEL_EXPORTER_OTLP[_signal]_HEADERS` family is never relayed onto gateway pods. Correct on two counts: Envoy can't read collector auth from env (it needs the header in exporter config, so relaying is inert), and it avoids spreading a collector credential onto gateway pods — collector auth here is transport-level (mesh mTLS). This is a security tightening over the prior approved revision.
- ✅ **Looks good:** Test coverage is added in the same commit (`envoy_test.go`): an injected `OTEL_EXPORTER_OTLP_HEADERS=Authorization=Bearer secret` is asserted **absent** from the rendered gateway env, alongside the existing `service.name` non-relay and `agent.id` identity assertions.
- ✅ **Looks good:** `logging.md` is updated in the same commit to document the transport-level-auth / headers-not-relayed behavior — code and architecture doc move together.

### Findings
No new defects this re-review. Carried-over, non-blocking observations from prior reviews (summary-only):

- 🟢 **Suggestion:** `gatewayOTelEnv` sets `OTEL_RESOURCE_ATTRIBUTES` fresh (`envoy.go:1529`), fully replacing the controller's relayed value. The `agent.id`/namespace override is right, but it also drops any platform-wide attributes the Operator may have parked in the inherited value (e.g. `k8s.cluster.name`). Consider merging rather than replacing if those matter downstream. (Accepted as-is in prior review.)
- 🟢 **Suggestion:** The exporter parse (`config.go:142`) keeps only host/port/scheme of `OTEL_EXPORTER_OTLP_ENDPOINT` and discards any path component; for OTLP/HTTP the per-signal URI is rebuilt as `scheme://host:port/v1/traces`, so a path-prefixed endpoint like `https://collector/otlp` would lose its `/otlp` prefix. Fine for the common host:port collector endpoint; worth a guard if path-prefixed endpoints are expected. (Accepted as-is in prior review.)

### Documentation Check (doc-drift)
**Verdict:** minor drift

**Drift:** none.

**Possible drift:**
- ⚠️ **Check 2 (`Last verified:` staleness)** — `docs/architecture/logging.md:3` reads `Last verified: 2026-06-24`. The page's "Gateway telemetry" section is accurate and was updated alongside the code in this same commit; the date matches the commit's *author* date (2026-06-24) but the branch was rebased so the committer/PR-HEAD date is now 2026-06-29. Marginal — the content is in sync; only the date string trails the rebase. _Human judgement:_ bump to the PR/HEAD date, or leave as the authoring date.

The "Gateway telemetry" section accurately reflects the rendered Envoy bootstrap, the relay model, the headers-not-relayed / transport-level-auth behavior, and the credential-redaction invariants; no ADR references leak in.

### Verdict
APPROVE — The new commit is a clean, well-tested security tightening (drop the inert-and-credential-bearing OTLP `*_HEADERS` relay) with the architecture doc updated in lockstep. Nothing new to block on; the two carried-over suggestions remain non-blocking and the only doc nit is a marginal `Last verified:` date trailing the rebase.

---
_Review by [DAM](https://github.com/dam-agents/dam) · automated code guardian_

**Review by dam-code-guardian — APPROVED** [review]
🛡️ **DAM** — ✅ Code Review @ `44e46a5`

## PR #1810: feat(controller): emit gateway Envoy telemetry from operator-relayed OTel env
**Author:** pilartomas | **Branch:** feat/gateway-otel-ready → main | **Changes:** +815 −9 (7 files)

### Summary
Makes the agent gateway (Envoy) OpenTelemetry-ready: the controller relays its own inherited `OTEL_*` env onto each gateway pod and renders native Envoy tracing / access-log / stats config in the bootstrap, off by default and credential-safe by design. This re-review covers a **rebase onto newer `main`** — the PR's own contribution is unchanged from the previously-approved revision (`1080095`); no gateway-telemetry logic moved. The only in-scope delta is that `docs/architecture/logging.md`'s `Last verified:` is now bumped to `2026-07-03`, which resolves the marginal date-trailing-the-rebase nit noted in the prior review. _(@willallstet's "docs reviewed — no update required" comment noted; the PR updates `logging.md` regardless. No findings suppressed; no PR-local overrides on file.)_

### Changes since last review
Previous HEAD: `1080095` (2026-06-29T09:42:54Z) — verdict APPROVE

- ✅ **Fixed:** `logging.md` `Last verified:` date now matches the current PR/HEAD date (`2026-07-03`) — the prior review's only doc nit (date trailing the rebase) is resolved.
- 🔁 **Still present:** `OTEL_RESOURCE_ATTRIBUTES` is set fresh on the gateway, fully replacing the controller's relayed value (`packages/controller/pkg/reconciler/envoy.go`) — drops any platform-wide attributes the Operator parked there. Non-blocking; accepted as-is in prior reviews.
- 🔁 **Still present:** `OTelExporter()` discards any path component of `OTEL_EXPORTER_OTLP_ENDPOINT` (`packages/controller/pkg/config/config.go:135`), so a path-prefixed OTLP/HTTP endpoint would lose its base path. Non-blocking; accepted as-is in prior reviews.

### Findings
No new defects this re-review (rebase only — the gateway-telemetry feature is byte-for-byte the previously-approved logic). Carried-over, non-blocking observations (summary-only):

- 🟢 **Suggestion:** `OTEL_RESOURCE_ATTRIBUTES` override drops inherited platform-wide attributes — consider merging rather than overwriting. (`packages/controller/pkg/reconciler/envoy.go`)
- 🟢 **Suggestion:** Path component of `OTEL_EXPORTER_OTLP_ENDPOINT` is discarded; add a guard or doc note if path-prefixed collector endpoints are ever in scope. (`packages/controller/pkg/config/config.go:135`)

### Documentation Check (doc-drift)
**Verdict: aligned**

✅ No documentation drift detected. `docs/architecture/logging.md` is the only architecture page touched; its "Gateway telemetry" section accurately reflects the rendered Envoy bootstrap (relay model, three signals, transport-level collector auth / `*_HEADERS` not relayed, and the four credential-redaction invariants), the `Last verified:` header is at the current PR date (`2026-07-03`), the two cross-references into `observability.md` (`#platform-service-export`, `#trusted-attribution`) both resolve to real headings, and no ADR references or volatile package/path content leak in.

### Verdict
APPROVE — Rebase-only re-review of an already-solid, well-tested, default-off telemetry wiring; the doc date nit from last time is fixed and the two carried-over suggestions remain non-blocking edge cases.

---
_Review by [DAM](https://github.com/dam-agents/dam) · automated code guardian_

**Review by dam-code-guardian — APPROVED** [review]
🛡️ **DAM** — ✅ Code Review @ `0d0152e`

## PR #1810: feat(controller): emit gateway Envoy telemetry from operator-relayed OTel env
**Author:** pilartomas | **Branch:** feat/gateway-otel-ready → main | **Changes:** +941 −9 (7 files)

### Summary
Makes the agent gateway (Envoy) OpenTelemetry-ready: the controller relays its own inherited `OTEL_*` env onto each gateway pod and renders native Envoy tracing / access-log / stats config in the bootstrap, off by default and credential-safe by design. This re-review covers one new commit since `44e46a5` — a genuine hardening pass (not a rebase). _(@willallstet's "docs reviewed — no update required" comment noted; the PR updates `logging.md` regardless. No findings suppressed; no PR-local overrides on file.)_

### Changes since last review
Previous HEAD: `44e46a5` (2026-07-03T06:26:25Z) — verdict APPROVE

- 🆕 **New — ✅:** Gateway resource-attribute identity renamed `agent.id` → `platform.gateway.id` (`packages/controller/pkg/reconciler/envoy.go:1755`). Security-positive: `agent.*` attributes are agent-forgeable through the transit chain and the collector sanitizes only `platform.agent.id`, so a `platform.*`-namespaced key survives untouched and stays trustworthy. Doc + all trace/comment references updated in lockstep, and `TestEnvoyContainer_RelaysOTelEnvWithGatewayIdentity` now asserts `platform.gateway.id=…`.
- 🆕 **New — ✅:** Collector-bound export CONNECTs get a dedicated zero-sampled route (`envoy.go` outer-listener match on `:authority` = `{{ .TelemetryCollectorHost }}:{{ .TelemetryCollectorPort }}`), so the pipeline never traces its own periodic telemetry pushes. Gated on `.Telemetry && $.OTel.Traces`; pinned by `TestRenderEnvoyBootstrap_CollectorConnectNotTraced` (and its negative case: no tracer → no exclusion route).
- 🆕 **New — ✅:** Error-only access log on the `terminate_otel_collector` transit chain (`envoy.go`), gated on `$.OTel.AccessLogs`, with an `or_filter` of `status_code_filter (>= 400)` + `response_flag_filter` so delivery failures become visible while steady-state success stays unlogged. Path rendered via `REQ_WITHOUT_QUERY` — consistent with the credential-redaction invariant. Pinned by `TestRenderEnvoyBootstrap_TransitChainErrorOnlyAccessLog`.
- 🆕 **New — ✅:** `otel_export` cluster `connect_timeout` 1s → 5s (`envoy.go`), matching the transit cluster and the tracer/sink request timeouts; sensible since export is background work and the budget now also covers the TLS handshake under `$.OTel.Secure`.
- 🆕 **New — ✅:** Test-anchoring fix — `value: 100` / `value: 10` sampling assertions replaced with `random_sampling`-anchored regexes, since a bare `value: 100` is a substring of the listener's `port_value: 10000` (a real false-positive risk in the prior assertions).
- 🔁 **Still present:** `OTEL_RESOURCE_ATTRIBUTES` is set fresh on the gateway, fully replacing the controller's relayed value (`packages/controller/pkg/reconciler/envoy.go:1757`) — drops any platform-wide attributes the Operator parked there. Non-blocking; accepted as-is in prior reviews.
- 🔁 **Still present:** `OTelExporter()` discards any path component of `OTEL_EXPORTER_OTLP_ENDPOINT` (`packages/controller/pkg/config/config.go:135`), so a path-prefixed OTLP/HTTP endpoint would lose its base path. Non-blocking; accepted as-is in prior reviews.

### Findings
No new defects this re-review — the new commit is clean, coherent, and each behavior change lands with a focused render test. Carried-over, non-blocking observations (summary-only):

- 🟢 **Suggestion:** `OTEL_RESOURCE_ATTRIBUTES` override drops inherited platform-wide attributes — consider merging rather than overwriting. (`packages/controller/pkg/reconciler/envoy.go`)
- 🟢 **Suggestion:** Path component of `OTEL_EXPORTER_OTLP_ENDPOINT` is discarded; add a guard or doc note if path-prefixed collector endpoints are ever in scope. (`packages/controller/pkg/config/config.go:135`)

### Documentation Check (doc-drift)
**Verdict: aligned**

✅ No documentation drift detected. `docs/architecture/logging.md` is the only architecture page touched; its "Gateway telemetry" section was updated in lockstep with this commit — the `agent.id` → `platform.gateway.id` rename is reflected in both the prose and the "Bounded cardinality" bullet (with the forgeability rationale), transport-level collector auth and `*_HEADERS`-not-relayed are described, `Last verified:` is at the PR date (`2026-07-03`), and the two cross-references into `observability.md` (`#platform-service-export`, `#trusted-attribution`) both resolve to real headings. The zero-sampled collector route and error-only transit-chain access log are Envoy-bootstrap implementation details consistent with the section's existing description — no drift. No ADR references or volatile package/path content leak in.

### Verdict
APPROVE — A well-scoped hardening commit: a security-motivated resource-attribute rename, self-tracing suppression, delivery-failure visibility, and a timeout alignment, each backed by a focused test (including a genuine false-positive fix in the sampling assertions). The two carried-over suggestions remain non-blocking edge cases.

---
_Review by [DAM](https://github.com/dam-agents/dam) · automated code guardian_

**Review by dam-code-guardian — APPROVED** [review]
🛡️ **DAM** — ✅ Code Review @ `bc2f646`

## PR #1810: feat(controller): emit gateway Envoy telemetry from operator-relayed OTel env
**Author:** pilartomas | **Branch:** feat/gateway-otel-ready → main | **Changes:** +1281 −9 (8 files)

### Summary
Makes the agent gateway (Envoy) OpenTelemetry-ready: the controller renders native Envoy tracing / access-log / stats config into each gateway's bootstrap, off by default and credential-safe by design. This re-review covers one new commit since `0d0152e` — a dedicated gateway exporter env pair (`PLATFORM_GATEWAY_OTLP_ENDPOINT`/`_PROTOCOL`) that decouples the gateway's exporter transport from the controller SDK's OTLP/HTTP-only env, plus OTLP-exported access logs on all four listener chains. _(@willallstet's "docs reviewed — no update required" comment noted; the PR updates `logging.md` regardless. No findings suppressed; no PR-local overrides on file.)_

### Changes since last review
Previous HEAD: `0d0152e` (2026-07-03T09:19:44Z) — verdict APPROVE

- 🆕 **New — ✅:** Dedicated gateway exporter knob (`config.go`) — `OTelExporter()` now prefers `PLATFORM_GATEWAY_OTLP_ENDPOINT`/`_PROTOCOL` over the relayed `OTEL_*`, extracted into a shared `parseOTLPExporter` helper. The chart sets it to the collector's gRPC `:4317` (`deployment.yaml`), gated under `clickstack.enabled`, so the gateway gets gRPC (stats-sink capable) while the controller SDK keeps its OTLP/HTTP `:4318` — the two consumers no longer constrain each other's transport. Backed by `TestOTelExporter_GatewayOverrideWins` (incl. the protocol-unset→gRPC default) and `TestRenderEnvoyBootstrap_GatewayOverrideDecouplesFromControllerEnv`.
- 🆕 **New — ✅:** OTLP access-log export added on all four chains (outer egress, terminate-chains, otel-collector transit error-log, L4 passthrough), each mirroring the existing stdout JSON record to `cluster: otel_export`. Redaction invariants preserved — `REQ_WITHOUT_QUERY(:PATH)` used throughout and no `Authorization` operator; the transit-chain OTLP twin keeps the error-only `or_filter` (status ≥ 400 + response-flag).
- 🆕 **New — ✅:** `gatewayOTelEnv` now states the _effective_ exporter in the pod env (override wins over relayed, gRPC default), so an override change rolls the pod like any other env change; `LogsURI` (`/v1/logs`) added to the view for the OTLP/HTTP branch.
- 🔁 **Still present:** `OTEL_RESOURCE_ATTRIBUTES` is set fresh on the gateway, fully replacing the controller's relayed value (`packages/controller/pkg/reconciler/envoy.go`) — drops any platform-wide attributes the Operator parked there. Non-blocking; accepted as-is in prior reviews.
- 🔁 **Still present:** `OTelExporter()`/`parseOTLPExporter` discards any path component of the OTLP endpoint (`packages/controller/pkg/config/config.go:135`), so a path-prefixed OTLP/HTTP endpoint would lose its base path. Non-blocking; accepted as-is in prior reviews.

### Findings
No new defects this re-review — the new commit is coherent and each behavior change lands with a focused render/config test. Carried-over, non-blocking observations (summary-only):

- 🟢 **Suggestion:** `OTEL_RESOURCE_ATTRIBUTES` override drops inherited platform-wide attributes — consider merging rather than overwriting. (`packages/controller/pkg/reconciler/envoy.go`)
- 🟢 **Suggestion:** Path component of the OTLP endpoint is discarded; add a guard or doc note if path-prefixed collector endpoints are ever in scope. (`packages/controller/pkg/config/config.go:135`)

### Documentation Check (doc-drift)
**Verdict: aligned**

✅ No documentation drift detected. `docs/architecture/logging.md` is the only architecture page touched; its "Gateway telemetry" section moves in lockstep with this commit — the dedicated `PLATFORM_GATEWAY_OTLP_*` exporter knob (gRPC, decoupled from the OTLP/HTTP-only SDK env, `OTEL_*` as BYO fallback) is described, access logs now note the OTLP export "beside every other platform service's logs", metrics are stated as gRPC-only with the explicit "BYO OTLP/HTTP collector gets no gateway stats" caveat, `Last verified:` is at the PR date (`2026-07-03`), the two cross-references into `observability.md` (`#platform-service-export`, `#trusted-attribution`) both resolve to real headings, and no ADR references or volatile package/path content leak in.

### Verdict
APPROVE — A well-scoped feature commit: a dedicated gRPC exporter env for the gateway (decoupled from the OTLP/HTTP-only controller SDK) plus OTLP-exported access logs, each backed by a focused test and with the credential-redaction invariants preserved; the architecture doc is updated in lockstep. The two carried-over suggestions remain non-blocking edge cases.

---
_Review by [DAM](https://github.com/dam-agents/dam) · automated code guardian_
